### ISSUE TITLE
typescript useDispatch error: Argument of type 'value' is not assignable to parameter of type 'never'

### ISSUE THREAD

**Issue by slugs99** [CharlesStover/reactn#94]
Hello, I'm out of luck trying to work out the typescript syntax and I couldn't find anything, helpful issues related to this.

`npx create-react-app reactn-ts --typescript && cd reactn-ts && npm i reactn`

```
// App.tsx

import React from "reactn";

const { addReducer, setGlobal, useDispatch, useGlobal } = React;

setGlobal({
  value: 2
});

addReducer("increment", (state: { value: number }, dispatch, action = {}) => {
  switch (action.type) {
    case "INC":
      return {
        value: state.value + action.payload
      };

    default:
      throw Error(`Invalid action type. Got type ${typeof action.type}`);
  }
});

interface G {
  value: number;
}

const App: React.FC = () => {
  const [value] = useGlobal<G>("value");
```

```
  /* error shows below this line */
  const dispatchIncrement = useDispatch("increment");
```

```
  const increment = (num: number) => {
    return dispatchIncrement({
      type: "INC",
      payload: num
    });
  };

  return (
    <div className="App">
      <p>value: {value}</p>
      <input type="button" value="INCREMENT" onClick={() => increment(2)} />
    </div>
  );
};

export default App;
```

P.S: The JS syntax of the code just works fine.

**Comment by CharlesStover** [comment]
Hi @slugs99 . Thanks for your engagement in the community.

I believe this error is due to ReactN not being able to infer the shape of your reducers/dispatchers. The same way you use `useGlobal<G>(...)` to tell it the shape of your global state, you would `useDispatch<G, R>(...)` to tell it the shape of your reducers.

If you are using the `addReducer` and `setGlobal` helper functions, I recommend adding a `global.d.ts` file that tracks your global state shape and reducer shape.

```
import 'reactn';

interface IncrementAction {
  payload: number;
  type: 'INC';
}

declare module 'reactn/default' {

  export interface Reducers {
    increment: (
      global: State,
      dispatch: Dispatch,
      action: IncrementAction,
    ) => Pick<State, 'value'>;
  }

  export interface State {
    value: number;
  }
}
```

Save this above content in `src/global.d.ts`, and you can use `useGlobal()` and `useDispatch()` without providing the `<G>` or `<G, R>` generics, as ReactN will use the definition in the aforementioned file. You can also remove the `interface G` from your `index.ts`.

**Comment by slugs99** [comment]
As a workaround `const dispatchIncrement = useDispatch().increment;` was simply working.

Now, I tried your suggestion but couldn't get it to work.

But the following steps worked:

1. Create `global.d.ts` as you suggested
2. `import { Reducers, State } from "reactn/default";`
3. 
```const [value] = useGlobal<State>("value");
const dispatchIncrement = useDispatch<State, Reducers>('increment');```

I think this is not what you are expecting, but I got an idea. This is okay for now, thank you
